rfcs/0032-github-merging-strategy.md

```diff
@@ -0,0 +1,134 @@
+minimise this because
+* PRs with conflicts do not merge automatically
+* the PR process is slowed down by waiting for these manual interventions
+* developer time is wasted in resolving these conflicts.
+
+Our current merge strategy is 'squash and merge', which discards the
+individually committed changes within each PR. This hampers `git`'s ability to
+reduce or eliminate merge conflicts by combining the changes from individual
+commits, unnecessarily creating more merge conflicts to be handled manually.
+
+These unnecessary conflicts are particularly prevelant in PR chains. For
+example,
+```
+PR 1
+   develop
+-> [Add line 1 in A.txt]
+-> [Modify line 1 in B.txt]
+
+PR 2 (chained to PR 1)
+   develop
+-> [Add line 1 in A.txt]
+-> [Modify line 1 in B.txt]
+-> [Modify line 1 in A.txt]
+-> ...
+```
+when `PR 1` is squashed and merged into develop, `git` is unable to recognise
+that the initial changes made in `PR 2` match with those from `PR 1` and
+reports a merge conflict.
+
+If PRs in chains have review comments that result in changes, these will
+usually also result in a merge conflict. For example,
+```
+PR 1
+   develop
+-> [Add line 1 in A.txt]
+-> [Modify line 1 in B.txt]
+-> [Modify line 1 in A.txt in response to review comments]
+
+PR 2 (chained to PR 1)
+   develop
+-> [Add line 1 in A.txt]
+-> [Modify line 1 in B.txt]
+-> [Add line 1 in C.txt]
+-> ...
+```
+when `PR 1` is squashed and merged, `git` will again be unable to merge `PR 2`
+automatically.
+
+## Detailed design
+[detailed-design]: #detailed-design
+
+We can use `git`'s merge commits when merging to avoid the above issues. These
+preserve the individual commit information that `git` needs to automatically
+resolve these conflicts.
+
+This is a simple configuration change in the GitHub settings for the repo.
+
+## Drawbacks
+[drawbacks]: #drawbacks
+
+#### We lose a linear `git` history (from `git log` and friends)
+
+We could mitigate this by running a branch in parallel with `develop` that
+'squashes' each merge commit after-the-fact with
+```bash
+git co develop-pretty
+git diff develop~ develop | git apply
+git commit -C develop
+```
+
+Locally, developers can also use `git log -m --first-parent` to view only the
+merge commits. `git log -m --first-parent --patch` shows the changes from
+commits beneath the merge commit, as expected.
+
+#### `git blame` identifies the commit within PRs instead of the merge commit
+
+Similar to the above, developers can locally use `git blame -m --first-parent`,
+or blame on the 'prettified' branch.
+
+#### The history will contain bad or useless commit messages
+
+This can be mostly ignored when using the above mitigations. Changing commit
+message culture is out of the scope of this RFC.
+
+## Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+* Why is this design the best in the space of possible designs?
+  - The alternatives (squashing and rebasing) both throw away the information
```

Comment:
Is it theoretically possible to auto-rebase `ready-to-merge-into-develop` branches onto their dependency branch's squashed merge commit into develop? For instance, observe the following workflow:

You have 2 PRs: A and B. B depends on the content in A, and contains the history of A along with a some extra new commits. Let's call the head of A that B is based off of A_h. When A is merged into develop, all new commits in A up to and including A_h are squashed into a single commit A_s. The bot which managed `ready-to-merge-into-develop` branches gets signaled that A was merged, and then goes to branch B, and rebases from A_h onto A_s. In most cases (so long as there were no additional conflicts that were dealt with on A only), this rebase should be clean of conflicts. If has conflicts, the bot can stop the process and leave a comment mentioning that the rebase should be done manually. If there were no conflicts, it will force push over B.

The main downside of this approach is that you could get your branch force-pushed from underneath you by the bot, while you are still working on it. Perhaps that trouble is too much to make this worthwhile. However, it doesn't seem that hard to teach people to just `git fetch -a && git rebase -i $(git show-ref -s origin/<branch>)` in order to fix this issue when it occurs (and it would only occur once per PR, and won't occur so long as you don't put the `ready-to-merge-into-develop` prematurely).

Reply:
> Is it theoretically possible to auto-rebase `ready-to-merge-into-develop` branches onto their dependency branch's squashed merge commit into develop?

Theoretically, sure. This only works for PR chains of length 2 though.

Consider `develop <- A <- B <- C` a chain of PRs. When `A` is squash-merged into `develop`, rebase `B` onto the squashed commit `A_s` from `A`. Now `C` is orphaned: its last common commit with both `B` and `develop` is `A_s~`. This puts us in effectively the same situation we have now with squash-to-merge for PR `C`.

Reply:
That's a good point. I guess you could get around this by chaining squashes all the way back to `C` immediately (I think), but that seems very error prone and too much effort to be worth it.